c++/src/kj/memory.h

```diff
@@ -168,6 +170,53 @@ class NullDisposer: public Disposer {
   void disposeImpl(void* pointer) const override {}
 };
 
+
+// =======================================================================================
+// Ptr Counters
+
+#ifdef KJ_DEBUG
+#define KJ_ASSERT_PTR_COUNTERS
+// When defined, keeps track of active Ptr<T> instances and asserts validity of their ownership
+#endif
+
+namespace _ {
+#ifdef KJ_ASSERT_PTR_COUNTERS
+class AtomicPtrCounter {
+// AtomicPtrCounter uses atomic operations to keep track of active pointers.
+// 
+// Thread safety notes:
+// - every dec() observes the effect of corresponding inc() because of release/consume order
+// - when assertEmpty() is called with balanced dec/inc calls, all releases were consumed, and
+//   additional load with consume memory order will see the correct last value.
+// - when assertEmpty() is called with unbalanced dec/inc calls int will observe the result of 
+//   one of the unbalanced inc(), which is always > 0.
+
+public:
+  void dec() {
+    size_t prevCount = count.fetch_sub(1, std::memory_order_consume);
+    KJ_IREQUIRE(prevCount > 0, "unbalanced inc/dec");
+  }
+
+  void inc() {
+    size_t prevCount = count.fetch_add(1, std::memory_order_release);
+    KJ_IREQUIRE(prevCount >= 0, "unbalanced inc/dec");
```

Comment:
This always passes since `prevCount` is unsigned. You could fix by doing `add_fetch` instead and then checking that the result is `> 0`.

But I think this check is not needed. `dec()` would have had to fail previously before this check could fail.